concourse/client/util.py

```diff
@@ -122,3 +156,163 @@ def jobs_not_triggered(
+            "pinned by technical user 'concourse', please do not unpin before "
+            f"{next_retry.isoformat()}"
+        ),
+    )
+    concourse_api.pin_and_comment_resource_version(
+        pipeline_name=resource.pipeline_name(),
+        resource_name=resource.name,
+        resource_version_id=resource_version.id(),
+        comment=json.dumps(dataclasses.asdict(pin_comment)),
+    )
+
+
+@contextmanager
+def ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    try:
+        _ensure_resource_unpinned(
+            resource=resource,
+            resource_version=resource_version,
+            concourse_api=concourse_api,
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+        info(f'successfully unpinned {resource.name=}')
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    previous_retry_time = None
+    maximal_unpin_checktime = datetime.now() + timedelta(minutes=5)
+
+    while maximal_unpin_checktime - datetime.now() >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        try:
+            pin_comment = from_dict(
+                data_class=PinComment,
+                data=json.loads(cc_resource.pin_comment()),
+            )
+            if pin_comment.version == resource_version.version():
+                raise PinningUnnecessaryError("Resource is already pinned for the same version")
+
+            next_retry_time = dateutil.parser.isoparse(pin_comment.next_retry)
+            if previous_retry_time and next_retry_time == previous_retry_time:
+                warning(
+                    f'found resource which was not unpinned '
+                    f'{resource.pipeline_name()=} {resource.name=}. Unpinning resource now'
+                )
+                concourse_api.unpin_resource(
+                    pipeline_name=resource.pipeline_name(),
+                    resource_name=resource.name,
+                )
+                return
+        except json.JSONDecodeError:
+            info(
+                f'Resource comment for {resource.name=} of {resource.pipeline_name()=} '
+                'could not be parsed - most likely pinned by human user'
+            )
+        except DaciteError:
+            info(
+                f'Resource comment for {resource.name=} of {resource.pipeline_name()=} '
+                'could not be instantiated'
+            )
+
+        previous_retry_time = next_retry_time
+        time.sleep((next_retry_time - datetime.now()).seconds + 5)
+
+    raise PinningFailedError(
+        f'Tried at least 5 minutes to unpin '
```

Comment:
use the actual value (time in minutes), rather than hard-coding.

also, rather use seconds?